src/Kestrel.Core/Internal/Http/Http1Connection.cs

```diff
@@ -413,67 +413,77 @@ protected override string CreateRequestId()
-                        case RequestProcessingStatus.ParsingHeaders:
-                            throw BadHttpRequestException.GetException(
-                                RequestRejectionReason.MalformedRequestInvalidHeaders);
-                    }
-                }
-                else if (!_keepAlive && _requestProcessingStatus == RequestProcessingStatus.RequestPending)
+            try
+            {
+                ParseRequest(result.Buffer, out consumed, out examined);
+            }
+            catch (InvalidOperationException)
+            {
+                if (_requestProcessingStatus == RequestProcessingStatus.ParsingHeaders)
                 {
-                    // Stop the request processing loop if the server is shutting down or there was a keep-alive timeout
-                    // and there is no ongoing request.
-                    return false;
+                    throw BadHttpRequestException.GetException(RequestRejectionReason
+                        .MalformedRequestInvalidHeaders);
                 }
-                else if (RequestTimedOut)
+                throw;
+            }
+            finally
+            {
+                Input.Advance(consumed, examined);
+            }
+
+            if (result.IsCompleted)
+            {
+                switch (_requestProcessingStatus)
                 {
-                    // In this case, there is an ongoing request but the start line/header parsing has timed out, so send
-                    // a 408 response.
-                    throw BadHttpRequestException.GetException(RequestRejectionReason.RequestTimeout);
+                    case RequestProcessingStatus.RequestPending:
+                        endConnection = true;
+                        return true;
+                    case RequestProcessingStatus.ParsingRequestLine:
+                        throw BadHttpRequestException.GetException(
+                            RequestRejectionReason.InvalidRequestLine);
+                    case RequestProcessingStatus.ParsingHeaders:
+                        throw BadHttpRequestException.GetException(
+                            RequestRejectionReason.MalformedRequestInvalidHeaders);
                 }
             }
+            else if (!_keepAlive && _requestProcessingStatus == RequestProcessingStatus.RequestPending)
+            {
+                // Stop the request processing loop if the server is shutting down or there was a keep-alive timeout
+                // and there is no ongoing request.
+                endConnection = true;
+                return true;
+            }
+            else if (RequestTimedOut)
+            {
+                // In this case, there is an ongoing request but the start line/header parsing has timed out, so send
+                // a 408 response.
+                throw BadHttpRequestException.GetException(RequestRejectionReason.RequestTimeout);
+            }
 
-            EnsureHostHeaderExists();
```

Comment:
Nit: Why not continue to call EnsureHostHeaderExists here if `_requestProcessingStatus == RequestProcessingStatus.AppStarted`? I think it cleans up HttpProtocol a bit to remove FinishParsing.

Reply:
Fair point. Will add that in.